astropy/coordinates/tests/test_representation.py

```diff
@@ -395,6 +419,26 @@ def test_getitem_scalar(self):
         with pytest.raises(TypeError):
             s_slc = s[0]
 
+    def test_transform(self):
+
+        s1 = UnitSphericalRepresentation(lon=[1, 2] * u.deg, lat=[3, 4] * u.deg)
+
+        matrix = rotation_matrix(-10, "z", u.deg)
+
+        s2 = s1.transform(matrix)
+
+        assert_allclose(s2.lon.deg, s1.lon.deg + 10)
+        assert_allclose(s2.lat.deg, s1.lat.deg)
+
+        # now with a non rotation matrix
+        matrix = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+
+        s3 = s1.transform(matrix)
+        expected = s1.to_cartesian().transform(matrix).represent_as(UnitSphericalRepresentation)
```

Comment:
Now that I see the test, I'm wondering about this case...  For straight up multiplication, we let the representation become spherical: 
https://github.com/astropy/astropy/blob/21f367a48b862f077bf00871dd7d9095c38eac73/astropy/coordinates/representation.py#L1594-L1597

(And I preserve that in #11470). I think similarly here the result should be a `SphericalRepresentation`, i.e., we should not just fall back to `UnitSpherical`. So, how about doing something like in #11470:
```
return self.dimensional_representation(...).transform(matrix)
```
Of course, you could then fall back to unitspherical if all the distances are equal to 1 -- or check for a rotation matrix after all....)

One advantage of that is that units are treated correctly as well.  Of course, for this test, it would just mean to do `represent_as(SphericalRepresentation)`.

Reply:
I'm implementing an O(3) check since improper rotations also preserve length, just not orientation.

```
atol = 1e-15
I = np.identity(matrix.shape[-1])
is_O3 = np.allclose(matrix @ matrix.swapaxes(-2, -1), I, atol=atol)
```

Reply:
Anything that satisfies O(3) remains unit-spherical. Everything else becomes spherical.

Reply:
@mhvk. Does this seem a good solution?

Reply:
@nstarman - the O(3) check looks good too, and in a way I liked in the original PR that we had this nice new utility function...  On the other hand, a test for all distances being close to unity is perhaps simpler and just as effective? Overall, I think I'd tend to go with the matrix check you propose...

Reply:
@mhvk . Re-added the utility function, better than before.

Reply:
@adrn @mhvk , I think ready for final review 🤞

Reply:
The current check is ``is_rotation()`` with ``allow_improper``. This is slightly more restrictive than the test I describe above, where any O(3) matrix stays in UnitSpherical. I'm adding an ``is_O3`` checker for things like reflections, which can also stay in UnitSpherical.
Also, the checks now use ``isclose`` to allow for checks *along* the axes. I moved the ``np.all()`` to ``transform``.